### PR TITLE
Add live single trigger fitting scripts

### PR DESCRIPTION
**Pull request by GarethDaviesGW**
Codes for fitting single-detector triggers for use in live singles FAR estimates

pycbc_single_trigger_fits is used (currently) to take a list of analysis dates in YYYY_MM_DD format, open a directory containing trigger files for that day, apply cuts and fit these to an exponential

pycbc_combine_single_trigger_fits takes these the output of pycbc_single_trigger_fits and combines the fits, giving a fit to all triggers (saved as mean), as well as a 95th percentile of the fit value (saved as conservative)

Both codes plot fits to the data, but these may be removed

These were not initially written for wider use, and so are not particularly readable/maintainable and may be too restrictive in places. This commit is just to get the files into version control. I will tidy, make input better, fail properly and give docstrings, helpful comments etc. later/soon

### CODE REVIEW

**Comment by titodalcanton** [comment]
I would like `live` to be explicit in the command names (not just in the script location in the repo) so as to avoid confusion with things like `pycbc_fit_sngl_trigs`. Can we have the paths be instead
```
bin/live/pycbc_live_combine_single_fits
bin/live/pycbc_live_single_trigger_fits
```
(hehe, [equal lengths](https://xkcd.com/276/) ;)

**Comment by titodalcanton** [comment]
And `live_utils` -> `live` is because eventually we might want to move `pycbc_live` itself there.

File: bin/live/pycbc_live_combine_single_fits [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+
+for f in files:
+    fits_f = h5py.File(f,'r')
+    if 'days_since_epoch' in fits_f.attrs:
+        analysis_dates += [fits_f.attrs['days_since_epoch']]
+    elif 'analysis_date' in fits_f.attrs:
+        analysis_year = int(fits_f.attrs['analysis_date'][:4])
+        analysis_month = int(fits_f.attrs['analysis_date'][5:7])
+        analysis_day = int(fits_f.attrs['analysis_date'][8:10])
+        n_since_epoch = (datetime.date(analysis_year, analysis_month, analysis_day)\
```
Comment (lpsinger) [inline bin/live/pycbc_live_combine_single_fits]: You don't need both parentheses _and_ an explicit line continuation.

File: bin/live/pycbc_live_single_trigger_fits [hunk 2]
```diff
@@ -0,0 +1,377 @@
...
+                                 ifo + '-' + args.analysis_date + "-TRIGGER-FITS.png")
+        plt.grid()
+        plt.xlim(args.fit_threshold, 1.05 * maxstat)
+        plt.ylim(0.5 / live_time[ifo] / max_ct, 1.5 * max_rate)
+        plt.xlabel(args.fit_param)
+        plt.ylabel("Number of louder triggers per live time per applicable template")
+        plt.legend()
+        logging.getLogger().setLevel(logging.WARNING)
+        plt.savefig(oput_plot)
+        logging.getLogger().setLevel(logging.INFO)
```
Comment (lpsinger) [inline bin/live/pycbc_live_single_trigger_fits]: Why are you changing the log level in the middle of the script?

File: bin/live/pycbc_live_combine_single_fits [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+        analysis_month = int(m.group(0)[5:7])
+        analysis_day = int(m.group(0)[8:10])
+        n_since_epoch = (datetime.date(analysis_year, analysis_month, analysis_day)\
+                         - datetime.date(2000,1,1)).days
+        analysis_dates += [n_since_epoch]
+    for ifo in args.ifos:
+        if ifo not in fits_f:
+            counts_all[ifo] += [-1 * np.ones_like(counts_all[ifo][-1])]
+            alphas_all[ifo] += [-1 * np.ones_like(alphas_all[ifo][-1])]
+            print(f + " has no " + ifo + " triggers")
```
Comment (lpsinger) [inline bin/live/pycbc_live_combine_single_fits]: Seems like you're using a mix of bare `print` statements and logging. Personally, I'd pick one or the other, and I'd use logging.

File: bin/live/pycbc_live_combine_single_fits [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+
+counts_bin = {ifo: [c for c in zip(*counts_all[ifo])] for ifo in args.ifos}
+alphas_bin = {ifo: [a for a in zip(*alphas_all[ifo])] for ifo in args.ifos}
+
+alphas_out = {ifo : np.zeros(len(alphas_bin[ifo])) for ifo in args.ifos}
+counts_out = {ifo : np.zeros(len(counts_bin[ifo])) for ifo in args.ifos}
+q05_alphas_out = {ifo : np.zeros(len(alphas_bin[ifo])) for ifo in args.ifos}
+q95_counts_out = {ifo : np.zeros(len(alphas_bin[ifo])) for ifo in args.ifos}
+
+fout = h5py.File(args.output_file,'w')
```
Comment (lpsinger) [inline bin/live/pycbc_live_combine_single_fits]: PEP 8: missing whitespace after comma.

File: bin/live/pycbc_live_single_trigger_fits (hunk 2 above)
Comment (GarethDaviesGW) [reply bin/live/pycbc_live_single_trigger_fits]: This was in order to stop the *incredibly* verbose logging output that savefig does given a logging.info level. I think Tito wrote something to get around this issue, so will look for it and see if I can implement that here

File: bin/live/pycbc_live_combine_single_fits [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+        mean_alpha = c.mean() / invalphan.mean()
+        q05_alpha = np.quantile(a, 0.05)
+        q05_alphas_out[ifo][counter] = q05_alpha
+        alphas_out[ifo][counter] = mean_alpha
+        q95_count = np.quantile(c, 0.95)
+        q95_counts_out[ifo][counter] = q95_count * len(c)
+        counts_out[ifo][counter] = c.sum()
+        if args.output_plot_file_format:
+            alpha_lines += ax_alpha.plot(ad[valid_alpha], a, c=colours[counter],
+                     label="duration {:.2f}-{:.2f}".format(l, u))
```
Comment (lpsinger) [inline bin/live/pycbc_live_combine_single_fits]: PEP8: ragged indentation

File: bin/live/pycbc_live_single_trigger_fits [hunk 6]
```diff
@@ -0,0 +1,377 @@
...
+
+# Also calculate live time so that this fitting can be used in rate estimation
+# Live time is not immediately obvious - get an approximation with 8 second
+# granularity by adding 8 seconds per 'valid' file
+
+live_time = {ifo: 0 for ifo in args.ifos}
+
+#store as a DictArray - nice as it has the .select() method
+events_dict = {k : [] for k in required_datasets}
+logging.info("Getting events from which meet criteria: {}".format(ineq_str))
```
Comment (lpsinger) [inline bin/live/pycbc_live_single_trigger_fits]: Logging methods do their own formatting, and are generally optimized to skip formatting operations if no logging handlers are configured. I'd suggest:

```python
logging.info("Getting events from which meet criteria: %s", ineq_str)
```

File: bin/live/pycbc_live_single_trigger_fits [hunk 7]
```diff
@@ -0,0 +1,377 @@
...
+        continue
+    with h5py.File(f,'r') as fin:
+        # Open the file: does it have the ifo group and snr dataset?
+        for ifo in args.ifos:
+            if not (ifo in fin and 'snr' in fin[ifo]):
+                continue
+            live_time[ifo] += 8
+            # all the required datasets must be present - this should
+            # not happen though as if snr is present then the others should be
+            if any([rd not in fin[ifo] for rd in required_datasets]):
```
Comment (lpsinger) [inline bin/live/pycbc_live_single_trigger_fits]: You can drop the square brackets; `any` can handle a generator and do short-circuit evaluation.

File: bin/live/pycbc_live_single_trigger_fits (hunk 2 above)
Comment (lpsinger) [reply bin/live/pycbc_live_single_trigger_fits]: So you can declare your own logger at module level and then configure the log level _just_ for logs that originate from it.

Comment (GarethDaviesGW) [reply bin/live/pycbc_live_single_trigger_fits]: Thanks! - I found that collin had written a PR for this exact reason, and we have a shortcut to do exactly that

Comment (GarethDaviesGW) [reply bin/live/pycbc_live_single_trigger_fits]: In the end I removed this plotting functionality anyway as it was using a custom template bank which makes it unsuitable for wider use

File: bin/live/pycbc_live_single_trigger_fits (hunk 6 above)
Comment (GarethDaviesGW) [reply bin/live/pycbc_live_single_trigger_fits]: That's good to know! Implemented here but I will look in other codes to see if that will improve efficiency

File: bin/live/pycbc_live_single_trigger_fits (hunk 7 above)
Comment (GarethDaviesGW) [reply bin/live/pycbc_live_single_trigger_fits]: Also a good thing to know - it's used elsewhere so I will fix those as well

**Pull request merged**